syntax/success_test.go

```diff
@@ -0,0 +1,234 @@
+package syntax
+
+import (
+	"bytes"
+	"encoding/hex"
+	"fmt"
+	"reflect"
+	"strings"
+	"testing"
+)
+
+func unhex(encoded string) []byte {
+	decoded, err := hex.DecodeString(encoded)
+	if err != nil {
+		panic(err)
+	}
+	return decoded
+}
+
+func buffer(size int) []byte {
+	return bytes.Repeat([]byte{0xA0}, size)
+}
+
+func hexBuffer(size int) string {
+	return strings.Repeat("A0", size)
+}
+
+type CrypticString string
+
+var (
+	crypticStringMarshalCalls   = 0
+	crypticStringUnmarshalCalls = 0
+)
+
+// A CrypticString marshalls as one length octet followed by the
+// UTF-8 bytes of the string, XOR'ed with an increasing sequence
+// starting with the length plus one (L+1, L+2, ...).
+func (cs CrypticString) MarshalTLS() ([]byte, error) {
+	crypticStringMarshalCalls += 1
+
+	l := byte(len(cs))
+	b := []byte(cs)
+	for i := range b {
+		b[i] ^= l + byte(i) + 1
+	}
+	return append([]byte{l}, b...), nil
+}
+
+func (cs *CrypticString) UnmarshalTLS(data []byte) (int, error) {
+	crypticStringUnmarshalCalls += 1
+
+	if len(data) == 0 {
+		return 0, fmt.Errorf("Length of CrypticString must be at least 1")
+	}
+
+	l := data[0]
+	if len(data) < int(l)+1 {
+		return 0, fmt.Errorf("TLS data not long enough for CrypticString")
+	}
+
+	b := data[1 : l+1]
+	for i := range b {
+		b[i] ^= l + byte(i) + 1
+	}
+
+	*cs = CrypticString(string(b))
+
+	return int(l + 1), nil
+}
+
+func TestSuccessCases(t *testing.T) {
+	dummyUint16 := uint16(0xB0A0)
+	testCases := map[string]struct {
+		value    interface{}
+		encoding []byte
+	}{
+		// Uints
+		"uint8": {
+			value:    uint8(0xA0),
+			encoding: unhex("A0"),
+		},
+		"uint16": {
+			value:    uint16(0xB0A0),
+			encoding: unhex("B0A0"),
+		},
+		"uint32": {
+			value:    uint32(0xD0C0B0A0),
+			encoding: unhex("D0C0B0A0"),
+		},
+		"uint64": {
+			value:    uint64(0xD0C0B0A090807060),
+			encoding: unhex("D0C0B0A090807060"),
+		},
+
+		// Varints
+		"varint8": {
+			value: struct {
+				V uint8 `tls:"varint"`
+			}{V: 0x3F},
+			encoding: unhex("3F"),
+		},
+		"varint16": {
+			value: struct {
+				V uint16 `tls:"varint"`
+			}{V: 0x3FFF},
+			encoding: unhex("7FFF"),
+		},
+		"varint32": {
+			value: struct {
+				V uint32 `tls:"varint"`
+			}{V: 0x3FFFFFFF},
+			encoding: unhex("BFFFFFFF"),
+		},
+		"varint64": {
+			value: struct {
+				V uint64 `tls:"varint"`
+			}{V: 0x3FFFFFFFFFFFFFFF},
+			encoding: unhex("FFFFFFFFFFFFFFFF"),
+		},
+
+		// Arrays
+		"array": {
+			value:    [5]uint16{0x1111, 0x2222, 0x3333, 0x4444, 0x5555},
+			encoding: unhex("11112222333344445555"),
```

Comment:
A test where the integers had different top and bottom bytes might be more intersting